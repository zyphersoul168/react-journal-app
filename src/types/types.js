export const types = {
  // authReducer
  login: '[auth] login',
  logout: '[auth] logout',

  // uiReducer
  uiSetError: '[UI] Set Error',
  uiRemoveError: '[UI] Remove Error',
  uiStartLoading: '[UI] Start loading',
  uiFinishLoading: '[UI] Finish loading',

  // notes
  notesAddNew: '[notes] New note',
  notesActive: '[notes] Set active note',
  notesLoad: '[notes] Load notes',
  notesUpdate: '[notes] Update note',
  notesFileURL: '[notes] Update img url',
  notesDelete: '[notes] Delete note',
  notesLogoutClean: '[notes] Logout clean',
};
