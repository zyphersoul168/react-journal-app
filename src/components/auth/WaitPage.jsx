import React from 'react';

export const WaitPage = () => {
  return (
    <div className="wait__container">
      <h1>{'Wait... :)'}</h1>
    </div>
  );
};
