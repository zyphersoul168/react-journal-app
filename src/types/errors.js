export const errors = {
  // firebase errors
  userNotFound: 'Firebase: Error (auth/user-not-found).',
  wrongPassword: 'Firebase: Error (auth/wrong-password).',
  existingAccount: 'Firebase: Error (auth/email-already-in-use).',
};
