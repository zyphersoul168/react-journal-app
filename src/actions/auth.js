import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/config';
import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { errors } from '../types/errors';
import { cleanNotes } from './notes';

const auth = getAuth();

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());
    try {
      const { user } = await signInWithEmailAndPassword(auth, email, password);
      dispatch(login(user.uid, user.displayName));
      dispatch(finishLoading());
    } catch (error) {
      dispatch(finishLoading());
      error.message === errors.userNotFound &&
        Swal.fire('Error', 'User not found', 'error');
      error.message === errors.wrongPassword &&
        Swal.fire('Error', 'Wrong Password', 'error');
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        error.message === errors.existingAccount &&
          Swal.fire('Error', 'There is already an account with this email', 'error');
        console.log(error.message);
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
    dispatch(cleanNotes());
  };
};

export const logout = () => ({
  type: types.logout,
});
