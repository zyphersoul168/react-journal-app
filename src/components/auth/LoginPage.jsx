import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginPage = () => {
  const dispatch = useDispatch();
  const { loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'ortiz@mail.com',
    password: 'aaroneduardo1',
  });

  const { email, password } = formValues;

  const handleLogin = (event) => {
    event.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <div>
      <h1>Log in</h1>
      <form onSubmit={handleLogin}>
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <button disabled={ loading } className="auth__button">
          <span className="auth__login">Log in</span>
        </button>
        <hr />
        <div>
          <span>Or you can log in with:</span>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Continue with google</b>
            </p>
          </div>
        </div>
        <span>
          Don't have an account?{' '}
          <Link className="link" to="/auth/register">
            Sign up
          </Link>
        </span>
      </form>
    </div>
  );
};
