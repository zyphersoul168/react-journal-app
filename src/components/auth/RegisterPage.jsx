import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterPage = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    name: 'aaron',
    email: 'ortiz@mail.com',
    password: '1234',
    password2: '1234',
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (event) => {
    event.preventDefault();
    isFormValid() && dispatch(startRegisterWithEmailPasswordName(email, password, name))
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError('Not valid email'));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(setError('Password too short'));
      return false;
    }

    dispatch(removeError());
    return true;
  };

  return (
    <div>
      <h1>Sign up</h1>
      <form onSubmit={handleRegister}>
        {msgError && <div className="auth__alert">{msgError}</div>}
        <input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="on"
          onChange={handleInputChange}
          value={name}
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          onChange={handleInputChange}
          value={email}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          onChange={handleInputChange}
          value={password}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          autoComplete="off"
          onChange={handleInputChange}
          value={password2}
        />
        <button className="auth__button">
          <span className="auth__login">Sign up</span>
        </button>
        <hr />
        <div>
          <span>Or you can sign up with:</span>
          <div className="google-btn">
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Continue with google</b>
            </p>
          </div>
        </div>
        <span>
          Already have an account?{' '}
          <Link className="link" to="/auth/login">
            Log in
          </Link>
        </span>
      </form>
    </div>
  );
};
