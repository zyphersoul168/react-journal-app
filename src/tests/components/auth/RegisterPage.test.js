import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

import { RegisterPage } from "../../../components/auth/RegisterPage";
import { types } from "../../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};
let store = mockStore(initialState);

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterPage />
    </MemoryRouter>
  </Provider>
);

describe("Tests on RegisterPage component", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch RegisterEmailPassword action", () => {
    const emailField = wrapper.find('input[name="email"]');
    emailField.simulate("change", {
      target: {
        value: "",
        name: "email",
      },
    });

    wrapper.find("form").simulate("submit", {
      preventDefault() {},
    });

    const actions = store.getActions();
    
    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: "Not valid email",
    });
  });

  test('should show alert message with error', () => {
    const initialState = {
      auth: {},
      ui: {
        loading: false,
        msgError: "Email no es correcto",
      },
    };
    const store = mockStore(initialState);
    
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterPage />
        </MemoryRouter>
      </Provider>
    );
  
    expect(wrapper.find(".auth__alert").exists()).toBe(true);
    expect(wrapper.find(".auth__alert").text().trim()).toBe(initialState.ui.msgError);
  });
  
});
