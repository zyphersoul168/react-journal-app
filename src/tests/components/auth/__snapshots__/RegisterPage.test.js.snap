// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on RegisterPage component should match snapshot 1`] = `
<div>
  <h1>
    Sign up
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="on"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="aaron"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="ortiz@mail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="1234"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="1234"
    />
    <button
      className="auth__button"
    >
      <span
        className="auth__login"
      >
        Sign up
      </span>
    </button>
    <hr />
    <div>
      <span>
        Or you can sign up with:
      </span>
      <div
        className="google-btn"
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Continue with google
          </b>
        </p>
      </div>
    </div>
    <span>
      Already have an account?
       
      <a
        className="link"
        href="/auth/login"
        onClick={[Function]}
      >
        Log in
      </a>
    </span>
  </form>
</div>
`;
